src/dbms/replication_handler.cpp

```diff
@@ -259,4 +277,101 @@ void RestoreReplication(replication::ReplicationState &repl_state, storage::Stor
       },
       repl_state.ReplicationData());
 }
+
+#ifdef MG_ENTERPRISE
+void CreateDatabaseHandler(DbmsHandler &dbms_handler, slk::Reader *req_reader, slk::Builder *res_builder) {
+  memgraph::storage::replication::CreateDatabaseReq req;
+  memgraph::slk::Load(&req, req_reader);
+
+  namespace sr = memgraph::storage::replication;
+  sr::CreateDatabaseRes res(sr::CreateDatabaseRes::Result::FAILURE);
+
+  // TODO
+  // Check epoch
+  // Check ts
+
+  try {
+    // Create new
+    auto new_db = dbms_handler.New(req.config);
+    if (new_db.HasValue()) {
+      // Successfully create db
+      res = sr::CreateDatabaseRes(sr::CreateDatabaseRes::Result::SUCCESS);
+    }
+  } catch (...) {
+    // Failure
+  }
+
+  memgraph::slk::Save(res, res_builder);
+}
+
+void SystemHeartbeatHandler(const uint64_t ts, slk::Reader *req_reader, slk::Builder *res_builder) {
+  replication::SystemHeartbeatReq req;
+  replication::SystemHeartbeatReq::Load(&req, req_reader);
+  memgraph::slk::Load(&req, req_reader);
+  replication::SystemHeartbeatRes res(ts);
+  memgraph::slk::Save(res, res_builder);
+}
+
+void SystemRecoveryHandler(DbmsHandler &dbms_handler, slk::Reader *req_reader, slk::Builder *res_builder) {
+  // TODO Speed up
+  memgraph::storage::replication::SystemRecoveryReq req;
+  memgraph::slk::Load(&req, req_reader);
+
+  namespace sr = memgraph::storage::replication;
+  sr::SystemRecoveryRes res(sr::SystemRecoveryRes::Result::FAILURE);
+
+  try {
+    // Get all current dbs
+    auto old = dbms_handler.All();
+    // Check/create the incoming dbs
+    for (const auto &config : req.database_configs) {
+      try {
+        dbms_handler.Get(config.uuid);
+      } catch (const UnknownDatabaseException &) {
+        // Missing db
+        if (dbms_handler.New(config).HasError()) {
+          spdlog::debug("Failed to create new database \"{}\".", config.name);
+          throw;
+        }
+      }
+      const auto it = std::find(old.begin(), old.end(), config.name);
+      if (it != old.end()) old.erase(it);
+    }
+    // Delete all the leftover old dbs
+    for (const auto &remove_db : old) {
+      if (dbms_handler.Delete(remove_db).HasError()) {
+        spdlog::debug("Failed to drop database \"{}\".", remove_db);
+        throw;
```

Comment:
throw what? Also:
DeleteError::NON_EXISTENT is fine. Something else removed it.
DeleteError::DEFAULT_DB is logic error. We shouldn't ever ask to drop default.

Reply:
Moved the slk save to OnExit, so I can just return.